doc/pano_modify.1
.\" Automatically generated by Pod::Man 2.22 (Pod::Simple 3.07)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "PANO_MODIFY ""1"""
.TH PANO_MODIFY "1" "2010-12-26" """Version: 2010.5.0""" "HUGIN"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
pano_modify \- Change output parameters of project file
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBpano_modify\fR [options] \fIinput.pto\fR
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBpano_modify\fR modifies a single Hugin .pto project.
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\fB\-o\fR, \fB\-\-output=file.pto\fR" 4
.IX Item "-o, --output=file.pto"
Output Hugin \s-1PTO\s0 file. Default: <filename>_mod.pto
.IP "\fB\-p\fR, \fB\-\-projection=NUMBER\fR" 4
.IX Item "-p, --projection=NUMBER"
Sets the output projection to number x
.IP "\fB\-\-fov=AUTO|HFOV|HFOVxVFOV\fR" 4
.IX Item "--fov=AUTO|HFOV|HFOVxVFOV"
Sets field of view.
.Sp
\&\s-1AUTO:\s0 calculates optimal fov
.Sp
HFOV|HFOVxVFOV: set to given fov
.IP "\fB\-s\fR, \fB\-\-straighten\fR" 4
.IX Item "-s, --straighten"
Straightens the panorama
.IP "\fB\-c\fR, \fB\-\-center\fR" 4
.IX Item "-c, --center"
Centers the panorama
.IP "\fB\-\-canvas=AUTO|WIDTHxHEIGHT\fR" 4
.IX Item "--canvas=AUTO|WIDTHxHEIGHT"
Sets the output canvas size
.Sp
\&\s-1AUTO:\s0 calculate optimal canvas size
.Sp
WIDTHxHEIGHT: set to given size
.IP "\fB\-\-crop=AUTO|left,right,top,bottom\fR" 4
.IX Item "--crop=AUTO|left,right,top,bottom"
Sets the crop rectangle
.Sp
\&\s-1AUTO:\s0 autocrop panorama
.Sp
left,right,top,bottom: to given size
.IP "\fB\-h\fR, \fB\-\-help\fR" 4
.IX Item "-h, --help"
Shows help
.SH "AUTHORS"
.IX Header "AUTHORS"
Thomas Modes


NEWS
Newsworthy changes in 0.7 beta 5
--------------------------------
* much improved vignetting and exposure correction,
  http://hugin.sf.net/tech
* ability to create HDR images during stitching
  (still extremly memory hungry, will be improved
   in the next version).
* fixed bad HFOV/VFOV limits of some projections
* OSX build should not crash all the time
* Added "Apply template" function, to apply the
  settings of a previously saved project to
  the current images.
* enabled "load lens data" button in dialog asking
  for focal length/HFOV.

Newsworthy changes in 0.7 beta 4
-------------------------
* update panotools to latest version
* Albers equal area conic and Miller cylindrical projection
* Support for straight line control points
* Fixed running of Autopano-SIFT integration on windows
* (linux) work around bugs in wxWidgets (fixes double deletion
  of control points when pressing the Del key)
* Force even values for width for 360 degree equirectangular
  panoramas.

Newsworthy changes in 0.7 beta 3
-------------------------
* Assistant panel, enables 3 click stitching (well, almost ;-)
* Center and rotate panorama by mouse clicks in preview window.
  Also added a numerical transform dialog
* automatic straighten
* support for Lambert and Lambert Azimuthal projections
* separate crop tab replaces buggy crop dialog
* crop is saved in lens .ini files
* Rotated display of images in control point selector, depending
  on the roll value
* allow images with and without alpha channels in the same project
* option to open final panorama in image editor after stitching
* present output of command line utils in nice window (unix)
* many small bugfixes
* removed image center dialog


Newsworthy changes in 0.6.1
---------------------------
* Autopano-SIFT integration was broken in 0.6 on Unix and OSX.
* Support for displaying HDR files within hugin, using log or gamma mapping
* .hdr file support (tif HDR files have been always useable)
* reduced memory usage, better image caching strategy in GUI.
* fixed crashes and hangs when using the preview auto update and avoid
  unnessecary remapping of images.
* vignetting correction estimation speed up.
* nona and nona_gui can be used to do template stitching, by specifying
  images on the command line (or dnd shortcut under windows)
* distribute fulla.exe with windows package (it was accidentally
  not included in the 0.6 package)
* universal build for Intel Macs, fixed hang after enblend or autopano-sift
  call
* projection can be changed in preview.
* removed all references to malicious panotools.info website
* various smaller bugfixes

Newsworthy changes in 0.6
-------------------------
* updated vigra to version 1.4. All sourcecode of hugin is now
  released under approved open source licences. This means there
  is no problem with including hugin in debian and ubuntu anymore.
* Copy ICC profiles from input images to output
* Added vignetting correction and brighness correction
  during stitching.
  Vignetting can be corrected by addition or division.
  The correction can be done with a flatfield image or an
  radial polynomial
* Functionality to estimate the vignetting polynomial
  directly from the overlapping images. This is an experimental feature.
* Multithreaded remapping. hugin and nona fully utilize multiple
  CPUs or multi core CPUs during preview generation and stitching
* support for dcraw 16 bit .ppm files
* Disabled PTOptimizer progress display on OSX. This should fix
  the crashes some people have experienced.
* New command line tool (fulla), for distortion, TCA and vignetting correction.
  Can use PTLens database (old format only).
* reorganized GUI layout.
* Czech translation added

Newsworthy changes in 0.5
-------------------------
* Added Catalan, Chinese and Hungarian translation
* Support TIFF compressions: NONE, DEFLATE, LZW
* option to create cropped tiff files for usage with enblend 2.4
* Mac OSX: fixed bug during calling PTOptimizer executable
* Windows only:
    hugin is shipped with the pano12.dll build with MSVC.
    This might lead to crashes when using PTStitcher.exe
    In that case replace pano12.dll with pano12_for_usage_with ptstitcher.dll

Newsworthy changes in 0.5 rc2
-----------------------------
* Removed all non-opensource sourcecode from CVS
* Fixed various crashes
* Japanese, Russian, Brazilian and Dutch translation
* Proper default arguments for Autopano-SIFT on Windows
  (users of hugin 0.5 r1 or less need to restore the default
   config)

Newsworthy changes in 0.5 rc1
-----------------------------
* Options on the menu and toolbar that are irrelevant
  are now disabled.
* enblend can now be called from nona even if it has 
  spaces in the path.

Newsworthy changes in 0.5 beta7
-------------------------------
* Updated translations
* Fixes to the build system on linux and windows

Newsworthy changes in 0.5 beta6
-------------------------------
* All files are now in source tarball
* French translation now complete

Newsworthy changes in 0.5 beta5
-------------------------------

* Support for modifying the size of the right/left panes in many tabs.
* All controls can now be reached on smaller displays as many areas 
  of the interface are now scrollable.
* All image control buttons in the preview tab can now be accessed, 
  allowing support for far more images.

* OSX build now supports unicode.

* Enblend arguments specified in preferences are now passed on correctly.
* Multiple tiff file outputs are no longer deleted incorrectly.
* Simultaneous scrolling in control points editor tabs respects borders properly.


src/tools/pano_modify.cpp
// -*- c-basic-offset: 4 -*-

/** @file pano_modify.cpp
 *
 *  @brief program to modify some aspects of a project file on the command line
 *  
 *  @author Thomas Modes
 *
 *  $Id$
 *
 */

/*  This program is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU General Public
 *  License as published by the Free Software Foundation; either
 *  version 2 of the License, or (at your option) any later version.
 *
 *  This software is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public
 *  License along with this software; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 */

#include <hugin_version.h>

#include <fstream>
#include <sstream>
#include <getopt.h>
#ifndef WIN32
 #include <unistd.h>
#endif
#include <panodata/Panorama.h>
#include <algorithms/nona/CenterHorizontally.h>
#include <algorithms/basic/StraightenPanorama.h>
#include <algorithms/nona/FitPanorama.h>
#include <algorithms/basic/CalculateOptimalScale.h>
#include <algorithms/basic/CalculateOptimalROI.h>

using namespace std;
using namespace HuginBase;
using namespace AppBase;

static void usage(const char * name)
{
    cout << name << ": change output parameters of project file" << endl
         << "pano_modify version " << DISPLAY_VERSION << endl
         << endl
         << "Usage:  " << name << " [options] input.pto" << endl
         << endl
         << "  Options:" << endl
         << "     -o, --output=file.pto  Output Hugin PTO file. Default: <filename>_mod.pto" << endl
         << "     -p, --projection=x     Sets the output projection to number x" << endl
         << "     --fov=AUTO|HFOV|HFOVxVFOV   Sets field of view" << endl
         << "                                   AUTO: calculates optimal fov" << endl
         << "                                   HFOV|HFOVxVFOV: set to given fov" << endl
         << "     -s, --straighten       Straightens the panorama" << endl
         << "     -c, --center           Centers the panorama" << endl
         << "     --canvas=AUTO|num%|WIDTHxHEIGHT  Sets the output canvas size" << endl
         << "                                   AUTO: calculate optimal canvas size" << endl
         << "                                   num%: scales the optimal size by given percent" << endl
         << "                                   WIDTHxHEIGHT: set to given size" << endl
         << "     --crop=AUTO|left,right,top,bottom  Sets the crop rectangle" << endl
         << "                                   AUTO: autocrop panorama" << endl
         << "                                   left,right,top,bottom: to given size" << endl
         << "     -h, --help             Shows this help" << endl
         << endl;
}

// toupper is overloaded in <locale>, but we want to use it as a unary function.
inline char toupper_2(char c)
{
    return toupper(c);
}

/** convert string to uppercase letters. */
std::string strToUpper(const std::string& aString)
{
    std::string result(aString);
    std::transform(aString.begin(), aString.end(), result.begin(), toupper_2);
    return result;
};

int main(int argc, char *argv[])
{
    // parse arguments
    const char * optstring = "o:p:sch";

    static struct option longOptions[] = {
        {"output", required_argument, NULL, 'o' },
        {"projection", required_argument, NULL, 'p' },
        {"fov", optional_argument, NULL, 'f' },
        {"straighten", no_argument, NULL, 's' },
        {"center", no_argument, NULL, 'c' },
        {"canvas", optional_argument, NULL, 'w' },
        {"crop", optional_argument, NULL, 'r' },
        {"help", no_argument, NULL, 'h' },
        0
    };

    int projection=-1;
    double newHFOV=-1;
    double newVFOV=-1;
    int scale=100;
    int newWidth=-1;
    int newHeight=-1;
    vigra::Rect2D newROI(0,0,0,0);
    bool doFit=false;
    bool doStraighten=false;
    bool doCenter=false;
    bool doOptimalSize=false;
    bool doAutocrop=false;
    int c;
    int optionIndex = 0;
    string output;
    string param;
    while ((c = getopt_long (argc, argv, optstring, longOptions,&optionIndex)) != -1)
    {
        switch (c) 
        {
            case 'o':
                output = optarg;
                break;
            case 'h':
                usage(argv[0]);
                return 0;
            case 'p':
                //projection
                projection=atoi(optarg);
                if((projection==0) && (strcmp(optarg,"0")!=0))
                {
                    cerr << "Could not parse projection number.";
                    return 1;
                };
                if(projection>=panoProjectionFormatCount())
                {
                    cerr << "projection " << projection << " is an invalid projection number.";
                    return 1;
                };
                break;
            case 'f':
                //field of view
                param=optarg;
                param=strToUpper(param);
                if(param=="AUTO")
                {
                    doFit=true;
                }
                else
                {
                    int hfov, vfov;
                    int n=sscanf(optarg, "%dx%d", &hfov, &vfov);
                    if(n==1)
                    {
                        if(hfov>0)
                        {
                            newHFOV=hfov;
                        }
                        else
                        {
                            cerr << "Invalid field of view" << endl;
                            return 1;
                        };
                    }
                    else
                    {
                        if (n==2)
                        {
                            if(hfov>0 && vfov>0)
                            {
                                newHFOV=hfov;
                                newVFOV=vfov;
                            }
                            else
                            {
                                cerr << "Invalid field of view" << endl;
                                return 1;
                            };
                        }
                        else
                        {
                            cerr << "Could not parse field of view" << endl;
                            return 1;
                        };
                    };
                };
                break;
            case 's':
                doStraighten=true;
                break;
            case 'c':
                doCenter=true;
                break;
            case 'w':
                //canvas size
                param=optarg;
                param=strToUpper(param);
                if(param=="AUTO")
                {
                    doOptimalSize=true;
                }
                else
                {
                    int pos=param.find("%");
                    if(pos!=string::npos)
                    {
                        param=param.substr(0,pos);
                        scale=atoi(param.c_str());
                        if(scale==0)
                        {
                            cerr << "No valid scale factor given." << endl;
                            return 1;
                        };
                        doOptimalSize=true;
                    }
                    else
                    {
                        int width, height;
                        int n=sscanf(optarg, "%dx%d", &width, &height);
                        if (n==2)
                        {
                            if(width>0 && height>0)
                            {
                                newWidth=width;
                                newHeight=height;
                            }
                            else
                            {
                                cerr << "Invalid canvas size" << endl;
                                return 1;
                            };
                        }
                        else
                        {
                            cerr << "Could not parse canvas size" << endl;
                            return 1;
                        };
                    };
                };
                break;
            case 'r':
                //crop
                param=optarg;
                param=strToUpper(param);
                if(param=="AUTO")
                {
                    doAutocrop=true;
                }
                else
                {
                    int left, right, top, bottom;
                    int n=sscanf(optarg, "%d,%d,%d,%d", &left, &right, &top, &bottom);
                    if (n==4)
                    {
                        if(right>left && bottom>top && left>=0 && top>=0)
                        {
                            newROI.setUpperLeft(vigra::Point2D(left,top));
                            newROI.setLowerRight(vigra::Point2D(right,bottom));
                        }
                        else
                        {
                            cerr << "Invalid crop area" << endl;
                            return 1;
                        };
                    }
                    else
                    {
                        cerr << "Could not parse crop values" << endl;
                        return 1;
                    };
                };
                break;
            case ':':
                cerr <<"Option " << longOptions[optionIndex].name << " requires a number" << endl;
                return 1;
                break;
            case '?':
                break;
            default:
                abort ();
        }
    }

    if (argc - optind != 1) 
    {
        cout << "Warning: pano_modify can only work on one project file at one time" << endl << endl;
        usage(argv[0]);
        return 1;
    };
    
    // set some options which depends on each other
    if(doStraighten)
    {
        doCenter=false;
        doFit=true;
    };
    if(doCenter)
    {
        doFit=true;
    };

    string input=argv[optind];
    // read panorama
    Panorama pano;
    ifstream prjfile(input.c_str());
    if (!prjfile.good()) {
        cerr << "could not open script : " << input << endl;
        return 1;
    }
    pano.setFilePrefix(hugin_utils::getPathPrefix(input));
    DocumentData::ReadWriteError err = pano.readData(prjfile);
    if (err != DocumentData::SUCCESSFUL) {
        cerr << "error while parsing panos tool script: " << input << endl;
        cerr << "DocumentData::ReadWriteError code: " << err << endl;
        return 1;
    }

    // sets the projection
    if(projection!=-1)
    {
        PanoramaOptions opt=pano.getOptions();
        opt.setProjection((PanoramaOptions::ProjectionFormat)projection);
        pano_projection_features proj;
        if (panoProjectionFeaturesQuery(projection, &proj)) 
            cout << "Setting projection to " << proj.name << endl;
        pano.setOptions(opt);
    };
    // straighten
    if(doStraighten)
    {
        cout << "Straighten panorama" << endl;
        StraightenPanorama(pano).run();
        CenterHorizontally(pano).run();
    };
    // center
    if(doCenter)
    {
        cout << "Center panorama" << endl;
        CenterHorizontally(pano).run();
    }
    //fit fov
    if(doFit)
    {
        cout << "Fit panorama field of view to best size" << endl;
        PanoramaOptions opt=pano.getOptions();
        CalculateFitPanorama fitPano = CalculateFitPanorama(pano);
        fitPano.run();
        opt.setHFOV(fitPano.getResultHorizontalFOV());
        opt.setHeight(roundi(fitPano.getResultHeight()));
        cout << "Setting field of view to " << opt.getHFOV() << " x " << opt.getVFOV() << endl;
        pano.setOptions(opt);
    };
    //set field of view manually
    if(newHFOV>0)
    {
        PanoramaOptions opt=pano.getOptions();
        opt.setHFOV(newHFOV);
        if(opt.fovCalcSupported(opt.getProjection()) && newVFOV>0)
            opt.setVFOV(newVFOV);
        cout << "Setting field of view to " << opt.getHFOV() << " x " << opt.getVFOV() << endl;
        pano.setOptions(opt);
    };
    // calc optimal size
    if(doOptimalSize)
    {
        cout << "Calculate optimal size of panorama" << endl;
        double s = CalculateOptimalScale::calcOptimalScale(pano);
        PanoramaOptions opt=pano.getOptions();
        opt.setWidth(roundi(opt.getWidth()*s*scale/100), true);
        cout << "Setting canvas size to " << opt.getWidth() << " x " << opt.getHeight() << endl;
        pano.setOptions(opt);
    };
    // set canvas size
    if(newWidth>0 && newHeight>0)
    {
        PanoramaOptions opt=pano.getOptions();
        opt.setWidth(newWidth);
        opt.setHeight(newHeight);
        cout << "Setting canvas size to " << opt.getWidth() << " x " << opt.getHeight() << endl;
        pano.setOptions(opt);
    };
    // auto crop
    if(doAutocrop)
    {
        cout << "Searching for best crop rectangle" << endl;

        CalculateOptimalROI cropPano(pano);
        cropPano.run();
        
        vigra::Rect2D roi=cropPano.getResultOptimalROI();
        PanoramaOptions opt = pano.getOptions();
        //set the ROI - fail if the right/bottom is zero, meaning all zero
        if(roi.right() != 0 && roi.bottom() != 0)
        {
            opt.setROI(roi);
            cout << "Set crop size to " << roi.left() << "," << roi.top() << "," << roi.right() << "," << roi.bottom() << endl;
            pano.setOptions(opt);
        }
        else
            cout << "Could not find best crop rectangle" << endl;
    };
    //setting crop rectangle manually
    if(newROI.right() != 0 && newROI.bottom() != 0)
    {
        PanoramaOptions opt = pano.getOptions();
        opt.setROI(newROI);
        cout << "Set crop size to " << newROI.left() << "," << newROI.right() << "," << newROI.top() << "," << newROI.bottom() << endl;
        pano.setOptions(opt);
    };

    //write output
    OptimizeVector optvec = pano.getOptimizeVector();
    UIntSet imgs;
    fill_set(imgs,0, pano.getNrOfImages()-1);
    // Set output .pto filename if not given
    if (output=="")
    {
        output=input.substr(0,input.length()-4).append("_mod.pto");
    }
    ofstream of(output.c_str());
    pano.printPanoramaScript(of, optvec, pano.getOptions(), imgs, false, hugin_utils::getPathPrefix(input));
    
    cout << endl << "Written output to " << output << endl;
    return 0;
}
